src/fairseq2/checkpoint.py

```diff
@@ -0,0 +1,329 @@
+            The training step number.
+        """
+
+    @abstractmethod
+    def load_last_checkpoint(self) -> Tuple[int, Dict[str, Any]]:
+        """Load the last checkpoint.
+
+        :returns:
+            - The training step number of the checkpoint.
+            - The checkpoint state.
+        """
+
+    @abstractmethod
+    def get_last_step_nr(self) -> Optional[int]:
+        """Return the training step number of the last checkpoint."""
+
+    @abstractmethod
+    def has_last_checkpoint(self) -> bool:
+        """Return ``True`` if there is a checkpoint marked as last."""
+
+    @abstractmethod
+    def save_consolidated_fsdp_model(self, step_nr: int, model: Module) -> None:
+        """Save ``model`` with a consolidated ``state_dict``.
+
+        This method consolidates the state of the model from all ranks so that
+        it can be loaded on a single device without FSDP.
+        """
+
+    @abstractmethod
+    def load_consolidated_model(
+        self, step_nr: int, model: Module, device: Optional[Device] = None
+    ) -> None:
+        """Load the consolidated model of the specified training step.
+
+        :param step_nr:
+            The training step number.
+        :param model:
+            The model to load.
+        :param device:
+            The device on which to load the model if it is on a meta device.
+        """
+
+    @abstractmethod
+    def load_last_consolidated_model(
+        self, model: Module, device: Optional[Device] = None
+    ) -> int:
+        """Load the last consolidated model.
+
+        :param model:
+            The model to load.
+        :param device:
+            The device on which to load the model if it is on a meta device.
+
+        :returns:
+            The training step number of the consolidated model.
+        """
+
+
+@final
+class FileCheckpointManager(CheckpointManager):
+    """Saves and loads training checkpoints on a file system."""
+
+    checkpoint_dir: Path
+    gang: Gang
+
+    def __init__(self, checkpoint_dir: Path, gang: Gang) -> None:
+        """
+        :param checkpoint_dir:
+            The root directory under which to store the checkpoints.
+        """
+        self.checkpoint_dir = checkpoint_dir
+        self.gang = gang
+
+    @finaloverride
+    def save_checkpoint(self, step_nr: int, state: Dict[str, Any]) -> None:
+        checkpoint = {"step_nr": step_nr, "data": state}
+
+        step_dir = self.checkpoint_dir.joinpath(f"step_{step_nr}")
+
+        if self.gang.rank == 0:
+            try:
+                step_dir.mkdir(parents=True, exist_ok=True)
+            except IOError as ex:
+                raise RuntimeError(
+                    f"The checkpoint directory for step {step_nr} cannot be created. See inner exception for details."
+                ) from ex
+
+        self.gang.barrier()
+
+        checkpoint_file = step_dir.joinpath(f"rank_{self.gang.rank}.pt")
+
+        try:
+            torch.save(checkpoint, checkpoint_file)
```

Comment:
For DDP, wouldn't we end up saving redundant copies of the model ?

Reply:
For `CheckpointManager` the provided `state` is an opaque dict. It does not really care what it contains. In a DDP setting, the trainer would do something like:

```
state_dict = {...}

if rank == 0:
    state_dict["model"] = model.state_dict()

checkpoint_manager.save_checkpoint(1234, state_dict)
```

Reply:
Okay so it is the responsibility of the caller to make sure they do not call `checkpoint_manager.save_checkpoint(1234, state_dict)` across multiple ranks when training with ddp.